backend/src/main/InMemoryDataSource.js

```diff
@@ -25,54 +37,103 @@ export default class InMemoryDataSource extends DataSource {
+        return context.jwtSign({ id: user.id });
+      }
+      if (duplicateEmail) {
+        return new Error('Email address already taken');
+      }
+      return new Error('Password is too short');
+    } catch (e) {
+      return e;
+    }
   }
 
-  createPost(data) {
+  loginUser(data, context) {
+    const { email, password } = data;
+    const user = this.users.find((u) => u.email === email);
+    const correctPassword = user && user.comparePassword(password);
+    if (user && correctPassword) {
+      return context.jwtSign({ id: user.id });
+    }
+    if (!user) {
+      return new Error('Unknown user');
+    }
+    return new Error('Wrong password');
+  }
+
+  createPost(data, context) {
     try {
-      const { name } = data.post.author;
-      const user = this.users.find((u) => u.name === name);
+      const { id: author } = context.user;
+      const user = this.users.find((u) => u.id === author);
       if (user) {
-        const post = new Post(data, this.getNewPostId());
+        const postData = {
+          post: {
+            title: data.post.title,
+            author: {
+              id: user.id,
+            },
+          },
+
+        };
+        const post = new Post(postData, InMemoryDataSource.getNewPostId());
         this.posts.push(post);
-        user.addPost(post.id);
         return post;
       }
+      return new Error('User not found');
     } catch (e) {
-      console.error(e);
-      return undefined;
+      return e;
     }
   }
 
-  deletePost({ id } = {}) {
-    try {
-      const post = this.posts.find((p) => p.id === id);
-      if (post) {
-        this.posts = this.posts.filter((p) => p.id !== post.id);
-        const user = this.users.find((u) => u.name === post.author);
-        if (user) {
-          user.removePost(post.id);
-        }
-        return post;
-      }
-    } catch (e) {
-      console.error(e);
-      return undefined;
+  deletePost({ id: postId } = {}, context) {
+    const { id: author } = context.user;
+    const user = this.users.find((u) => u.id === author);
+    const post = this.posts.find((p) => p.id === postId);
+    if (user && post && post.author === author) {
+      this.posts = this.posts.filter((p) => p.id !== post.id);
+      return post;
+    }
+    if (!user) {
+      return new Error('User not found');
+    }
+    if (!post) {
+      return new Error('Post not found');
     }
+    return new Error('User does not own the rights to delete the post');
   }
 
-  upvotePost({ id, voter = {} } = {}) {
-    const post = this.posts.find((p) => p.id === id);
-    return post && post.upvote(voter.name);
+  upvotePost({ id: postId } = {}, context) {
+    const { id: author } = context.user;
+    const user = this.users.find((u) => u.id === author);
+    const post = this.posts.find((p) => p.id === postId);
+    if (user && post) {
+      return post.upvote(author);
+    }
+    if (!user) {
+      return new Error('User not found');
+    }
```

Comment:
```suggestion
    if (!user) throw new Error('User not found');
```

Move that further up and make it a proper guard clause! Code execution won't reach any code after `return` or `throw`.